### PR TITLE
Implement ConnectionState and closeOnExit=graceful/always/never

### LINKED ISSUE #2563: If the default shell can't be launched and closeOnExit=true, Terminal buys the farm
# Environment

```
Windows build number:  10.0.18362.295
Windows Terminal version: 0.4.2382.0
```

# Steps to reproduce

* Install a WSL distribution (openSUSE Leap 15.0 in my case).
* Move it to top spot and make it default.
* Uninstall old WSL distribution and install the upgraded one (openSUSE Leap 15.1 in my case).

# Expected behavior

Remove uninstalled WSL distributions. Add all installed WSL distributions unless explicitly disabled in WT.

# Actual behavior

WT closes right away. I'm somehow expected to know the UUID change (`wsl.exe --list` does not show the UUIDs).

----
Another bug report claims that behavior for non-existent profiles would be handled gracefully in 0.4. Can't confirm that.

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
## Summary of the Pull Request

This pull request implements the new `ITerminalConnection::ConnectionState` interface (enum, event) and connects it through TerminalControl to Pane, Tab and App as specified in #2039. It does so to implement `closeOnExit` = `graceful` in addition to the other two normal CoE types.

It also:

* ... exposes the singleton `CascadiaSettings` through a function that looks it up by using the current Xaml application's `AppLogic`.
    * In so doing, we've broken up the weird runaround where App tells TerminalSettings to CloseOnExit and then later another part of App _asks TerminalControl_ to tell it what TerminalSettings said App told it earlier. `:crazy_eyes:`
* ... wires up a bunch of connection state points to `AzureConnection`. This required moving the Azure connection's state machine to use another enum name (oops).
* ... ships a helper class for managing connection state transitions.
* ... contains a bunch of template magic.
* ... contains a TODO (whoops, sorry, if you point it out you'll be removed from the selfhost program)
* ... introduces `WINRT_CALLBACK`, a oneshot callback like `TYPED_EVENT`.
* ... replaces a bunch of disparate `_connecting` and `_closing` members with just one uberstate.

## References

Specification is in PR #2039.

## PR Checklist
* [x] Closes #2563
* [x] CLA
* [ ] Tests added/passed
* [x] Requires documentation to be updated
* [x] I've discussed this with core contributors already.

### CODE REVIEW

File: doc/cascadia/profiles.schema.json [hunk 1]
```diff
@@ -325,9 +325,21 @@
             "type": "string"
           },
           "closeOnExit": {
-            "default": true,
-            "description": "When set to true (default), the selected tab closes when the connected application exits. When set to false, the tab will remain open when the connected application exits.",
-            "type": "boolean"
+            "default": "graceful",
+            "description": "When set to graceful/true (default), the selected tab closes when the connected application exits successfully. When set to always, the tab will never be closed automatically. When set to never/false, the tab will remain open when the connected application exits.",
```
Comment (greg904) [inline doc/cascadia/profiles.schema.json]: "When set to always, the tab will never be closed automatically."

isn't this reversed?

Comment (DHowett-MSFT) [reply doc/cascadia/profiles.schema.json]: It absolutely is. Good catch.

File: src/cascadia/TerminalConnection/ConptyConnection.h [hunk 2]
```diff
@@ -14,37 +17,31 @@ namespace wil
...
         void Resize(uint32_t rows, uint32_t columns);
         void Close();
 
         winrt::guid Guid() const noexcept;
 
+        WINRT_CALLBACK(TerminalOutput, TerminalOutputHandler);
+
     private:
         HRESULT _LaunchAttachedClient() noexcept;
+        void _indicateExitWithStatus(unsigned int status) noexcept;
```
Comment (carlos-zamora) [inline src/cascadia/TerminalConnection/ConptyConnection.h]: Nit: pascal case to match other formats

File: doc/cascadia/profiles.schema.json [hunk 3]
```diff
@@ -325,9 +325,21 @@
             "type": "string"
           },
           "closeOnExit": {
-            "default": true,
```
[the remark this answers is not in the document]
Comment (DHowett-MSFT) [reply doc/cascadia/profiles.schema.json]: It already does; that's the `oneOf` bit below.

File: src/cascadia/TerminalConnection/Resources/en-US/Resources.resw [hunk 4]
```diff
@@ -205,4 +205,13 @@
...
     <comment>The user must enter this character to signify NO</comment>
   </data>
+  <data name="ProcessExited" xml:space="preserve">
+    <value>[process exited with code %u]</value>
+    <comment>The first argument (%u) is the (positive) error code of the process. 0 is success.</comment>
+  </data>
+  <data name="ProcessFailedToLaunch" xml:space="preserve">
+    <value>[error 0x%8.08x when launching `%ls']</value>
+    <comment>The first argument is the hexadecimal error code. The second is the process name.
+If this resource spans multiple lines, it will not be displayed properly. Yeah.</comment>
```
[the remark this answers is not in the document]
Comment (DHowett-MSFT) [reply src/cascadia/TerminalConnection/Resources/en-US/Resources.resw]: We can hack around it, but I am not sure we can _fix_ it.

[the remark this answers is not in the document]
Comment (DHowett-MSFT) [reply src/cascadia/TerminalConnection/Resources/en-US/Resources.resw]: (newlines get compiled to `\n` by the resource compiler. If you manually add `\r\n` (through their HTML entities `&#x0D;...`) they get stripped by the resource compiler.)

[the remark this answers is not in the document]
Comment (DHowett-MSFT) [reply src/cascadia/TerminalConnection/Resources/en-US/Resources.resw]: (the terminal needs both `\r` _and_ `\n` to move down a line.

File: src/cascadia/inc/cppwinrt_utils.h [hunk 5]
```diff
@@ -71,6 +71,20 @@ public:
 private:                                                                                                                                           \
     winrt::event<winrt::Windows::Foundation::TypedEventHandler<sender, args>> _##name##Handlers;
 
+// This is a helper macro for both declaring the signature of a callback (nee event) and
```
[the remark this answers is not in the document]
Comment (DHowett-MSFT) [reply src/cascadia/inc/cppwinrt_utils.h]: `nee` like `used to be named`

File: src/cascadia/TerminalConnection/ConptyConnection.h (hunk 2 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalConnection/ConptyConnection.h]: I'm trying to get all new code to comply with #3512, actually!

File: src/cascadia/LocalTests_TerminalApp/SettingsTests.cpp [hunk 6]
```diff
@@ -1414,4 +1416,27 @@ namespace TerminalAppLocalTests
         auto terminalSettings = settings._profiles[0].CreateTerminalSettings(globalSettings.GetColorSchemes());
         VERIFY_ARE_EQUAL(expectedPath, terminalSettings.BackgroundImage());
     }
+    void SettingsTests::TestCloseOnExitCompatibilityShim()
```
[the remark this answers is not in the document]
Comment (DHowett-MSFT) [reply src/cascadia/LocalTests_TerminalApp/SettingsTests.cpp]: bahhhhhhhhhhhh

File: src/cascadia/TerminalConnection/EchoConnection.h [hunk 7]
```diff
@@ -5,23 +5,23 @@
 
 #include "EchoConnection.g.h"
 
+#include "../cascadia/inc/cppwinrt_utils.h"
```
Comment (zadjii-msft) [inline src/cascadia/TerminalConnection/EchoConnection.h]: at what point should this just be precomp'd?

File: src/cascadia/TerminalConnection/ConnectionStateHolder.h [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        // Return Value:
+        //   Whether we're in one of the states.
+        template<typename... Args>
+        bool _isStateOneOf(Args&&... args) const noexcept
+        {
+            static_assert((... && std::is_same<Args, ConnectionState>::value), "all queried connection states must be from the ConnectionState enum");
+            std::lock_guard<std::mutex> stateLock{ _stateMutex };
+            // dark magic: this is a C++17 fold expression that expands into
+            // state == 1 || state == 2 || state == 3 ...
+            return (... || (_connectionState == args));
```
Comment (zadjii-msft) [inline src/cascadia/TerminalConnection/ConnectionStateHolder.h]: I still don't understand what this dark magic does. is it equivalent to the following:

```c++
_connectionState == args[0] || _connectionState == args[1] || _connectionState == args[2] ...
```

(if you could index into the params like that?)

File: src/cascadia/TerminalApp/CascadiaSettings.cpp [hunk 9]
```diff
@@ -26,6 +27,21 @@ static constexpr std::wstring_view PACKAGED_PROFILE_ICON_PATH{ L"ms-appx:///Prof
 static constexpr std::wstring_view PACKAGED_PROFILE_ICON_EXTENSION{ L".png" };
 static constexpr std::wstring_view DEFAULT_LINUX_ICON_GUID{ L"{9acb9455-ca41-5af7-950f-6bca1bc9722f}" };
 
+// Method Description:
+// - Returns the settings currently in use by the entire Terminal application.
+// Throws:
+// - HR E_INVALIDARG if the app isn't up and running.
+const CascadiaSettings& CascadiaSettings::GetCurrentAppSettings()
+{
+    auto currentXamlApp{ winrt::Windows::UI::Xaml::Application::Current().as<winrt::TerminalApp::App>() };
```
[the remark this answers is not in the document]
Comment (DHowett-MSFT) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: Any time before AppLogic has LoadSettings'd. In our current lifecycle design, that is _never_ once general code in TerminalAPp is running.

File: src/cascadia/inc/cppwinrt_utils.h [hunk 10]
```diff
@@ -71,6 +71,20 @@ public:
 private:                                                                                                                                           \
     winrt::event<winrt::Windows::Foundation::TypedEventHandler<sender, args>> _##name##Handlers;
 
+// This is a helper macro for both declaring the signature of a callback (nee event) and
+// defining the body. Winrt callbacks need a method for adding a delegate to the
+// callback and removing the delegate. This macro will both declare the method
+// signatures and define them both for you, because they don't really vary from
+// event to event.
+// Use this in a class's header if you have a "delegate" type in your IDL.
+#define WINRT_CALLBACK(name, args)                                                           \
```
[the remark this answers is not in the document]
Comment (DHowett-MSFT) [reply src/cascadia/inc/cppwinrt_utils.h]: Yep, and I wanted to unify around the language Raymond Chen uses for WinRT things

File: src/cascadia/TerminalConnection/EchoConnection.h (hunk 7 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalConnection/EchoConnection.h]: Good question :)

File: src/cascadia/TerminalConnection/ConnectionStateHolder.h (hunk 8 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalConnection/ConnectionStateHolder.h]: YES! Exactly.

And the one above it is like

`typeof(args[0]) == ConnectionState && typeof(args[1] == ConnectionState) && ...`

File: src/cascadia/TerminalConnection/AzureConnection.cpp [hunk 11]
```diff
@@ -370,15 +334,15 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
                         }
                         catch (...)
                         {
-                            // Websocket has been closed
-                            if (!_closing.load())
+                            // Websocket has been closed; consider it a graceful exit?
+                            // This should result in our termination.
+                            if (_transitionToState(ConnectionState::Closed))
                             {
-                                _state = State::NoConnect;
```
Comment (zadjii-msft) [inline src/cascadia/TerminalConnection/AzureConnection.cpp]: Should we just remove the `AzureState::NoConnect` state entirely now?

File: src/cascadia/TerminalConnection/ConptyConnection.cpp [hunk 12]
```diff
@@ -205,32 +201,64 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
...
-        _disconnectHandlers();
+        winrt::hstring failureText{ wil::str_printf<std::wstring>(RS_(L"ProcessFailedToLaunch").c_str(), static_cast<unsigned int>(hr), _commandline.c_str()) };
+        _TerminalOutputHandlers(failureText);
+        _transitionToState(ConnectionState::Failed);
+
+        // Tear down any state we may have accumulated.
+        _hPC.reset();
+    }
+
+    void ConptyConnection::_indicateExitWithStatus(unsigned int status) noexcept
```
Comment (zadjii-msft) [inline src/cascadia/TerminalConnection/ConptyConnection.cpp]: doc comment

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 13]
```diff
@@ -156,8 +156,11 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
         // DON'T CALL _InitializeTerminal here - wait until the swap chain is loaded to do that.
 
         // Subscribe to the connection's disconnected event and call our connection closed handlers.
-        _connection.TerminalDisconnected([=]() {
-            _connectionClosedHandlers();
+        _connectionStateChangedRevoker = _connection.StateChanged(winrt::auto_revoke, [weakThis = get_weak()](auto&& /*s*/, auto&& /*v*/) {
```
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/TermControl.cpp]: This looks like a pattern we should probably be using in other places, right? Where the event handler doesn't capture a strong ref to the handler object? (maybe we should file a CodeHealth task for this)

Comment (DHowett-MSFT) [reply src/cascadia/TerminalControl/TermControl.cpp]: Yes, absolutely.

File: src/cascadia/TerminalConnection/ConptyConnection.cpp (hunk 12 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalConnection/ConptyConnection.cpp]: Thanks.

File: src/cascadia/TerminalConnection/AzureConnection.cpp (hunk 11 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalConnection/AzureConnection.cpp]: I'm working on the AZ conn in my free time, I'll defer this to then.

Comment (DHowett-MSFT) [reply src/cascadia/TerminalConnection/AzureConnection.cpp]: er defer this to that work

File: src/cascadia/TerminalConnection/Resources/en-US/Resources.resw (hunk 4 above)
[the remark this answers is not in the document]
Comment (DHowett-MSFT) [reply src/cascadia/TerminalConnection/Resources/en-US/Resources.resw]: 780 will likely be involved in this somehow :smile:

File: src/cascadia/TerminalConnection/ConnectionStateHolder.h (hunk 8 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalConnection/ConnectionStateHolder.h]: I'm going to expand these since i have to make a comment revision anyway

**Comment by mu88** [comment]
I curiously tried the new `graceful` behavior and one question came up to me. I've updated to `v0.8.10091.0` and have the following profile:
```
{
    "name" : "Build Script",
    "closeOnExit": "graceful",
    "guid" : "{65869F08-7651-44A5-9EAB-982A1577BC65}",
    "commandline" : "cmd.exe /k Batch\\Build.bat",
    "startingDirectory" : "C:\\Projects\\Code",
    "fontFace" : "Fira Code",
    "fontSize" : 10,
    "icon" : "C:\\Projects\\Code\\DotNet\\icon.ico"
}
```

When opening a new tab with *Build Script*, the script `Build.bat` gets triggered and succeeds. When executing `echo Exit Code is %errorlevel%` in the opened tab, `0` is returned. Nonetheless, the tab remains open. I would have expected the tab to close. Is my assumption wrong or is there any issue?

**Comment by DHowett-MSFT** [comment]
_if I recall correctly,_ You want `cmd /c`. `/k` means “continue running after you evaluate the command.” You can pretend it means “[k]eep running”.

**Comment by mu88** [comment]
You are absolutely right, thank you!

**Pull request merged**